pkg/virt-controller/watch/vm.go

```diff
@@ -1648,31 +1648,75 @@ func (c *VMController) enqueueVm(obj interface{}) {
 	c.Queue.Add(key)
 }
 
-func (c *VMController) removeVMIFinalizer(vmi *virtv1.VirtualMachineInstance) error {
-	vmiCopy := vmi.DeepCopy()
-	controller.RemoveFinalizer(vmiCopy, virtv1.VirtualMachineControllerFinalizer)
+func (c *VMController) getPatchFinalizerOps(oldObj, newObj v1.Object) ([]string, error) {
+	var ops []string
 
-	if equality.Semantic.DeepEqual(vmi.Finalizers, vmiCopy.Finalizers) {
-		return nil
+	oldFinalizers, err := json.Marshal(oldObj.GetFinalizers())
+	if err != nil {
+		return ops, err
 	}
 
-	log.Log.V(3).Object(vmi).Infof("VMI is in a final state. Removing VM controller finalizer")
-
-	ops := []string{}
-	newFinalizers, err := json.Marshal(vmiCopy.Finalizers)
+	newFinalizers, err := json.Marshal(newObj.GetFinalizers())
 	if err != nil {
-		return err
+		return ops, err
+	}
+
+	ops = append(ops, fmt.Sprintf(`{ "op": "test", "path": "/metadata/finalizers", "value": %s }`, string(oldFinalizers)))
+	ops = append(ops, fmt.Sprintf(`{ "op": "replace", "path": "/metadata/finalizers", "value": %s }`, string(newFinalizers)))
+	return ops, nil
+}
+
+func (c *VMController) removeVMIFinalizer(vmi *virtv1.VirtualMachineInstance) error {
+	if !controller.HasFinalizer(vmi, virtv1.VirtualMachineControllerFinalizer) {
+		return nil
 	}
-	oldFinalizers, err := json.Marshal(vmi.Finalizers)
+
+	log.Log.V(3).Object(vmi).Infof("VMI is in a final state. Removing VM controller finalizer")
+	newVmi := vmi.DeepCopy()
+	controller.RemoveFinalizer(newVmi, virtv1.VirtualMachineControllerFinalizer)
+	ops, err := c.getPatchFinalizerOps(vmi, newVmi)
 	if err != nil {
 		return err
 	}
-	ops = append(ops, fmt.Sprintf(`{ "op": "test", "path": "/metadata/finalizers", "value": %s }`, string(oldFinalizers)))
-	ops = append(ops, fmt.Sprintf(`{ "op": "replace", "path": "/metadata/finalizers", "value": %s }`, string(newFinalizers)))
+
 	_, err = c.clientset.VirtualMachineInstance(vmi.Namespace).Patch(context.Background(), vmi.Name, types.JSONPatchType, controller.GeneratePatchBytes(ops), &v1.PatchOptions{})
 	return err
 }
 
+func (c *VMController) removeVMFinalizer(vm *virtv1.VirtualMachine, finalizer string) (*virtv1.VirtualMachine, error) {
+	if !controller.HasFinalizer(vm, finalizer) {
+		return vm, nil
+	}
+
+	log.Log.V(3).Object(vm).Infof("VM is in a final state. Removing VM controller finalizer")
```

Comment:
This seems to be a general function but this log seems to be very specific. Maybe you want to include the finalizer we are removing?

Reply:
Leftover from initial implementation: Thanks!